modules/transport-grpc/src/main/java/org/opensearch/transport/grpc/GrpcPlugin.java

```diff
@@ -363,6 +336,48 @@ public Collection<Object> createComponents(
     ) {
         this.client = client;
 
+        // Initialize the interceptor chain with ThreadContext
+        this.serverInterceptor = new GrpcInterceptorChain(threadPool.getThreadContext());
+
+        List<OrderedGrpcInterceptor> orderedList = new ArrayList<>();
+
+        // Then add plugin-provided interceptors
+        // Similar to how ActionPlugin.getRestHandlerWrapper receives ThreadContext
+        if (!interceptorProviders.isEmpty()) {
+            for (GrpcInterceptorProvider provider : interceptorProviders) {
+                orderedList.addAll(provider.getOrderedGrpcInterceptors(threadPool.getThreadContext()));
+            }
+
+            // Validate that no two interceptors have the same order
+            Map<Integer, List<OrderedGrpcInterceptor>> orderMap = new HashMap<>();
+            for (OrderedGrpcInterceptor interceptor : orderedList) {
+                int order = interceptor.order();
+                orderMap.computeIfAbsent(order, k -> new ArrayList<>()).add(interceptor);
+            }
+
+            // Check for duplicates and throw exception if found
+            for (Map.Entry<Integer, List<OrderedGrpcInterceptor>> entry : orderMap.entrySet()) {
+                if (entry.getValue().size() > 1) {
+                    throw new IllegalArgumentException(
```

Comment:
I'm wondering if it would be helpful to output `entry.getValue()` in a friendly way (maybe call `getClass()` on each one?) to show which specific interceptors collide.

For the record, `ActionFilter` seems to allow duplicate order, but will (I think) just end up sorting things based on the plugin loading order. I think it's a good idea to be more strict with Grpc interceptors.